### PR TITLE
GEODE-3894: Create HandshakeRequest

### PR DESCRIPTION
**Pull request by WireBaron**
added connection state processors to inject protocol state specific
functionality refactored existing authentication and authorization flows to use
connection state processors added new handshake message, handler, and state
processor

Signed-off-by: Bruce Schuchardt <bschuchardt@pivotal.io>

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [x] Is your initial contribution a single, squashed commit?

- [x] Does `gradlew build` run cleanly?

- [x] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

@kohlmu-pivotal @galen-pivotal @PivotalSarge @bschuchardt 

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-client-protocol/src/main/java/org/apache/geode/internal/protocol/ConnectionShiroAuthorizingStateProcessor.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+package org.apache.geode.internal.protocol;
+
+import org.apache.geode.internal.security.SecurityService;
+import org.apache.geode.security.NotAuthorizedException;
+import org.apache.shiro.subject.Subject;
+import org.apache.shiro.util.ThreadState;
+
+public class ConnectionShiroAuthorizingStateProcessor implements ConnectionStateProcessor {
+  SecurityService securityService;
+  Subject subject;
```
Comment (PivotalSarge) [inline geode-client-protocol/src/main/java/org/apache/geode/internal/protocol/ConnectionShiroAuthorizingStateProcessor.java]: Could `securityService` and `subject` be `final`?

File: geode-protobuf/src/main/java/org/apache/geode/internal/protocol/protobuf/ProtobufConnectionHandshakeStateProcessor.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+import org.apache.geode.internal.protocol.NoSecurityConnectionStateProcessor;
+import org.apache.geode.internal.protocol.OperationContext;
+import org.apache.geode.internal.protocol.ProtocolErrorCode;
+import org.apache.geode.internal.protocol.StateValidationFailedException;
+import org.apache.geode.internal.protocol.protobuf.operations.HandshakeRequestOperationHandler;
+import org.apache.geode.internal.security.SecurityService;
+
+public class ProtobufConnectionHandshakeStateProcessor
+    implements ConnectionHandshakingStateProcessor {
+  SecurityService securityService;
```
Comment (PivotalSarge) [inline geode-protobuf/src/main/java/org/apache/geode/internal/protocol/protobuf/ProtobufConnectionHandshakeStateProcessor.java]: Also perhaps `final`?

File: geode-protobuf/src/main/proto/connection_API.proto [hunk 3]
```diff
@@ -16,10 +16,30 @@
...
+}
+
 message AuthenticationRequest {
     map<string,string> credentials = 1;
 }
 
 message AuthenticationResponse {
     bool authenticated = 1;
-}
+}
```
Comment (PivotalSarge) [inline geode-protobuf/src/main/proto/connection_API.proto]: Maybe restore the newline?

File: geode-client-protocol/src/main/java/org/apache/geode/internal/protocol/ConnectionShiroAuthorizingStateProcessor.java (hunk 1 above)
Comment (WireBaron) [reply geode-client-protocol/src/main/java/org/apache/geode/internal/protocol/ConnectionShiroAuthorizingStateProcessor.java]: done

File: geode-protobuf/src/main/java/org/apache/geode/internal/protocol/protobuf/ProtobufConnectionHandshakeStateProcessor.java (hunk 2 above)
Comment (WireBaron) [reply geode-protobuf/src/main/java/org/apache/geode/internal/protocol/protobuf/ProtobufConnectionHandshakeStateProcessor.java]: done

File: geode-client-protocol/src/main/java/org/apache/geode/internal/protocol/ConnectionShiroAuthorizingStateProcessor.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+ */
+package org.apache.geode.internal.protocol;
+
+import org.apache.geode.internal.security.SecurityService;
+import org.apache.geode.security.NotAuthorizedException;
+import org.apache.shiro.subject.Subject;
+import org.apache.shiro.util.ThreadState;
+
+public class ConnectionShiroAuthorizingStateProcessor implements ConnectionStateProcessor {
+  SecurityService securityService;
```
[the remark this answers is not in the document]
Comment (WireBaron) [reply geode-client-protocol/src/main/java/org/apache/geode/internal/protocol/ConnectionShiroAuthorizingStateProcessor.java]: done

**Comment by WireBaron** [comment]
Pushed a new change that has the state processors moved to a better location, exception renamed to be more generic, and state processor variables made static final.

File: geode-protobuf/src/main/proto/connection_API.proto [hunk 5]
```diff
@@ -16,6 +16,26 @@
 syntax = "proto3";
 package org.apache.geode.internal.protocol.protobuf;
 
+enum MajorVersions {
```
Comment (galen-pivotal) [inline geode-protobuf/src/main/proto/connection_API.proto]: A comment explaining what these are about would be nice: that current version will track the version that's deployed in the API. Could we write a test for this? Do we archive proto files somewhere? (maybe a problem for down the line)

File: geode-protobuf/src/test/java/org/apache/geode/internal/protocol/protobuf/operations/HandshakeRequestOperationHandlerJUnitTest.java [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+@Category(UnitTest.class)
+public class HandshakeRequestOperationHandlerJUnitTest {
+  private static final int INVALID_MAJOR_VERSION = 67;
```
Comment (galen-pivotal) [inline geode-protobuf/src/test/java/org/apache/geode/internal/protocol/protobuf/operations/HandshakeRequestOperationHandlerJUnitTest.java]: We should test against the defined constants (which we know are zero) `ConnectionAPI.MajorVersions.INVALID_MAJOR_VERSION` and `ConnectionAPI.MinorVersions.INVALID_MINOR_VERSION`. Perhaps we could parameterize some existing tests?

File: geode-protobuf/src/main/java/org/apache/geode/internal/protocol/protobuf/registry/ProtobufOperationContextRegistry.java [hunk 7]
```diff
@@ -109,5 +111,12 @@ private void addContexts() {
             opsResp -> ClientProtocol.Response.newBuilder().setGetAvailableServersResponse(opsResp),
             new ResourcePermission(ResourcePermission.Resource.CLUSTER,
                 ResourcePermission.Operation.READ)));
+
+    operationContexts.put(RequestAPICase.HANDSHAKEREQUEST,
+        new ProtobufOperationContext<>(ClientProtocol.Request::getHandshakeRequest,
+            new HandshakeRequestOperationHandler(),
+            opsResp -> ClientProtocol.Response.newBuilder().setHandshakeResponse(opsResp),
+            new ResourcePermission(ResourcePermission.Resource.DATA,
+                ResourcePermission.Operation.READ)));
```
Comment (galen-pivotal) [inline geode-protobuf/src/main/java/org/apache/geode/internal/protocol/protobuf/registry/ProtobufOperationContextRegistry.java]: I don't like that this is never used and is bypassed by the state processor. Perhaps we could leave a comment to that effect? Maybe mark it "ALL" here?

File: geode-client-protocol/src/main/java/org/apache/geode/internal/protocol/state/ConnectionAuthenticatingStateProcessor.java [hunk 8]
```diff
@@ -12,10 +12,18 @@
...
+package org.apache.geode.internal.protocol.state;
 
-import org.apache.geode.security.ResourcePermission;
+import org.apache.geode.internal.protocol.state.exception.ConnectionStateException;
 
-public interface Authorizer {
-  boolean authorize(Object authenticatedToken, ResourcePermission permissionRequested);
+import java.util.Properties;
+
+public interface ConnectionAuthenticatingStateProcessor extends ConnectionStateProcessor {
```
[the remark this answers is not in the document]
Comment (WireBaron) [reply geode-client-protocol/src/main/java/org/apache/geode/internal/protocol/state/ConnectionAuthenticatingStateProcessor.java]: That is correct, I added those comments.

File: geode-protobuf/src/main/java/org/apache/geode/internal/protocol/protobuf/ProtobufLocatorPipeline.java [hunk 9]
```diff
@@ -23,28 +23,32 @@
...
   private final ProtobufStreamProcessor streamProcessor;
+  private final ConnectionStateProcessor locatorConnectionState;
 
   ProtobufLocatorPipeline(ProtobufStreamProcessor protobufStreamProcessor,
       ProtocolClientStatistics statistics, InternalLocator locator) {
     this.streamProcessor = protobufStreamProcessor;
     this.statistics = statistics;
     this.locator = locator;
     this.statistics.clientConnected();
+    this.locatorConnectionState = new NoSecurityConnectionStateProcessor();
```
[the remark this answers is not in the document]
Comment (WireBaron) [reply geode-protobuf/src/main/java/org/apache/geode/internal/protocol/protobuf/ProtobufLocatorPipeline.java]: Yeah, would be nice if we also properly implemented security on the cache...

File: geode-protobuf/src/main/java/org/apache/geode/internal/protocol/protobuf/registry/ProtobufOperationContextRegistry.java (hunk 7 above)
Comment (WireBaron) [reply geode-protobuf/src/main/java/org/apache/geode/internal/protocol/protobuf/registry/ProtobufOperationContextRegistry.java]: That sounds like a great plan, but unfortunately there isn't an ALL option for resource or operation fields in ResourcePermissions.

File: geode-protobuf/src/main/proto/connection_API.proto (hunk 5 above)
Comment (WireBaron) [reply geode-protobuf/src/main/proto/connection_API.proto]: The real definitions of major and minor version are a little nebulous at that point, so I don't want to try to commit to anything at this point.

We do use the CURRENT_{MAJOR,MINOR}_VERSION throughout our tests, though we don't currently have any idea of other versions or how we might handle them (aside from outright failing version mismatches).

File: geode-protobuf/src/test/java/org/apache/geode/internal/protocol/protobuf/operations/HandshakeRequestOperationHandlerJUnitTest.java (hunk 6 above)
Comment (WireBaron) [reply geode-protobuf/src/test/java/org/apache/geode/internal/protocol/protobuf/operations/HandshakeRequestOperationHandlerJUnitTest.java]: Ugh, I hate that protobuf requires us to include 0 values for enums.  Added tests for these in the invalid version tests.

**Pull request merged**